src/websockethandler.cpp

```diff
@@ -0,0 +1,241 @@
+    close();
+  }
+}
+
+/**
+ * @brief Opens the websocket connection.
+ *
+ * No-op in case the connection is already open.
+ */
+void WebSocketHandler::open() {
+  if (m_webSocket.state() != QAbstractSocket::UnconnectedState &&
+      m_webSocket.state() != QAbstractSocket::ClosingState) {
+    logger.debug()
+        << "Attempted to open a WebSocket connection, but it's already open.";
+    return;
+  }
+
+  logger.debug() << "Attempting to open WebSocket connection."
+                 << WebSocketHandler::webSocketServerUrl();
+
+  QNetworkRequest request;
+  request.setRawHeader("Authorization",
+                       SettingsHolder::instance()->token().toLocal8Bit());
+  request.setUrl(QUrl(WebSocketHandler::webSocketServerUrl()));
+  m_webSocket.open(request);
+}
+
+/**
+ * @brief Ackowledges the WebSocket has been succesfully connected.
+ */
+void WebSocketHandler::onConnected() {
+  logger.debug() << "WebSocket connected";
+
+  connect(&m_webSocket, &QWebSocket::textMessageReceived, this,
+          &WebSocketHandler::onMessageReceived);
+
+  sendPing();
+}
+
+/**
+ * @brief Closes the websocket connection.
+ *
+ * No-op in case the connection is already closed.
+ */
+void WebSocketHandler::close() {
+  if (m_webSocket.state() == QAbstractSocket::UnconnectedState ||
+      m_webSocket.state() == QAbstractSocket::ClosingState) {
+    logger.debug()
+        << "Attempted to close a WebSocket, but it's already closed.";
+    return;
+  }
+
+  logger.debug() << "Closing WebSocket";
+  m_webSocket.close();
+}
+
+/**
+ * @brief Ackowledges the WebSocket has been closed.
+ *
+ * May trigger a reconnection attempt depending on whether the user is
+ * authenticated.
+ */
+void WebSocketHandler::onClose() {
+  logger.debug() << "WebSocket closed";
+
+  if (WebSocketHandler::isUserAuthenticated()) {
+    logger.debug()
+        << "User is authenticated. Attempting to reopen WebSocket in:"
+        << m_retryInterval;
+
+    QTimer::singleShot(m_retryInterval, this, &WebSocketHandler::open);
+  } else {
+    logger.debug()
+        << "User is not authenticated. Will not attempt to reopen WebSocket.";
+  }
+}
+
+/**
+ * @brief Sends a ping to the WebSocket server.
+ *
+ * WebSocket connections may silently be broken. Periodically pinging the server
+ * keeps track of the connection. Everytime a new ping is sent, a timer is
+ * started. If a response is received before the timer is up, the connection is
+ * alive and a new ping is scheduled. Otherwise the WebSocket is closed a a
+ * reconnection attempt is scheduled.
+ */
+void WebSocketHandler::sendPing() {
+  m_webSocket.ping();
+
+  m_pingTimer.setSingleShot(true);
+  m_pingTimer.start(m_pingInterval);
+}
+
+/**
+ * @brief Acknowledges a pong response from the WebSocket server. Schedules the
+ * next ping.
+ *
+ * @param elapsedTime How long it took to get the response since a ping was sent
+ * out.
+ */
+void WebSocketHandler::onPong(quint64 elapsedTime) {
+  qDebug() << "PONG PONG PONG";
```

Comment:
logger.debug() ?

Reply:
I was using qDebug because the logger.debug statements do not show up in the test logs for some reason.